prov/sm2/src/sm2_av.c

```diff
@@ -45,155 +46,104 @@ static int sm2_av_close(struct fid *fid)
-			ret = -FI_ENOMEM;
-		}
-
-		FI_INFO(&sm2_prov, FI_LOG_AV, "fi_addr: %" PRIu64 "\n", util_addr);
-		if (fi_addr)
-			fi_addr[i] = util_addr;
-
-		if (ret) {
-			if (util_av->eq)
-				ofi_av_write_event(util_av, i, -ret, context);
-			if (shm_id >= 0)
-				sm2_map_del(sm2_av->sm2_map, shm_id);
+		if (ret)
 			continue;
+
+		if (flags & FI_AV_USER_ID) {
+			sm2_av->sm2_aux[util_addr].cqfid = fi_addr[i];
+			FI_INFO(&sm2_prov, FI_LOG_AV,
+				"fi_addr: %d, USER_ID: %" PRIu64 "\n", util_addr,
+				fi_addr[i]);
 		}
 		else {
-			assert(shm_id >= 0 && shm_id < SM2_MAX_PEERS);
-			if (flags & FI_AV_USER_ID) {
-				assert(fi_addr);
-				sm2_av->sm2_map->peers[shm_id].fiaddr = fi_addr[i];
-			}
-			else {
-				sm2_av->sm2_map->peers[shm_id].fiaddr = util_addr;
-			}
-			succ_count++;
-			sm2_av->used++;
+			sm2_av->sm2_aux[util_addr].cqfid = util_addr;
+			FI_INFO(&sm2_prov, FI_LOG_AV, "fi_addr: %d\n", util_addr);
 		}
 
-		assert(sm2_av->sm2_map->num_peers > 0);
-
-		dlist_foreach(&util_av->ep_list, av_entry)
-		{
-			util_ep = container_of(av_entry, struct util_ep, av_entry);
-			sm2_ep = container_of(util_ep, struct sm2_ep, util_ep);
-			sm2_map_to_endpoint(sm2_ep->region, shm_id);
-			sm2_ep->region->max_sar_buf_per_peer =
-			    SM2_MAX_PEERS / sm2_av->sm2_map->num_peers;
+		if (fi_addr) {
+			fi_addr[i] = util_addr;
 		}
-	}
 
-	if (!(flags & FI_EVENT))
-		return succ_count;
+		succ_count++;
+	}
+	/* release the lock */
+	sm2_coordinator_unlock(&sm2_av->sm2_mmap);
 
-	ofi_av_write_event(util_av, succ_count, 0, context);
-	return 0;
+	if (flags & FI_EVENT) {
+		ofi_av_write_event(util_av, succ_count, 0, context);
+	}
+	return succ_count;
 }
 
 static int sm2_av_remove(struct fid_av *av_fid, fi_addr_t *fi_addr, size_t count,
 			 uint64_t flags)
 {
-	struct util_av *util_av;
-	struct util_ep *util_ep;
-	struct sm2_av *sm2_av;
-	struct sm2_ep *sm2_ep;
-	struct dlist_entry *av_entry;
-	int i, ret = 0;
-	int64_t id;
-
-	util_av = container_of(av_fid, struct util_av, av_fid);
-	sm2_av = container_of(util_av, struct sm2_av, util_av);
-
-	ofi_mutex_lock(&util_av->lock);
-	for (i = 0; i < count; i++) {
-		FI_INFO(&sm2_prov, FI_LOG_AV, "%" PRIu64 "\n", fi_addr[i]);
-		id = sm2_addr_lookup(util_av, fi_addr[i]);
-		ret = ofi_av_remove_addr(util_av, fi_addr[i]);
-		if (ret) {
-			FI_WARN(&sm2_prov, FI_LOG_AV,
-				"Unable to remove address from AV\n");
-			break;
-		}
-
-		sm2_map_del(sm2_av->sm2_map, id);
-		dlist_foreach(&util_av->ep_list, av_entry)
-		{
-			util_ep = container_of(av_entry, struct util_ep, av_entry);
-			sm2_ep = container_of(util_ep, struct sm2_ep, util_ep);
-			sm2_unmap_from_endpoint(sm2_ep->region, id);
-			if (sm2_av->sm2_map->num_peers > 0)
-				sm2_ep->region->max_sar_buf_per_peer =
-				    SM2_MAX_PEERS / sm2_av->sm2_map->num_peers;
-			else
-				sm2_ep->region->max_sar_buf_per_peer = SM2_BUF_BATCH_MAX;
-		}
-		sm2_av->used--;
-	}
-
-	ofi_mutex_unlock(&util_av->lock);
-	return ret;
+	/* No-OP */
+	return 0;
```

Comment:
This should not be a no-op. We need to remove it from the util_av (ie the private_aux for what you're doing right now).

Reply:
done